File: javascript/difference-of-squares/difference-of-squares.js
export class Squares {
  constructor(number) {
      this._number = number;
  }

  get sumOfSquares() {
        return (this._number * (this._number + 1) * (2 * this._number + 1)) / 6;
  }

  get squareOfSum() {
      return [...Array(this._number).keys()].reduce((accumulator, currentValue) => { return accumulator + (currentValue + 1)}, 0) ** 2;
  }

  get difference() {
      return this.squareOfSum - this.sumOfSquares;
  }
}

File: clojure/difference-of-squares/src/difference_of_squares.clj
(ns difference-of-squares)

(defn square [num]
  (* num num))

(defn sum-of-squares [num] 
  (/ (* num (inc num) (inc (* 2 num))) 6))

  ;; (->> (range 1 (inc num))
  ;;   (map #(* % %))
  ;;   (reduce +)))

  ;; (loop [n num
  ;;        acc 0]
  ;;   (if (<= n 0)
  ;;     acc
  ;;     (recur (dec n) (+ acc (* n n))))))

;; (defn sum-of-squares [num] 
;;   (if (<= num 0)
;;     0
;;     (+ (square num) (sum-of-squares (dec num)))))

(defn square-of-sum [num]
  (let [sum (reduce + (range 1 (inc num)))]
  (square sum)))

(defn difference [num]
  (- (square-of-sum num) (sum-of-squares num)))

